JIT 5: code specialization

jfrohnhofen opened the pull request:
ready for review now

I addressed all the feedback I got so far and added more documentation and comments to give a high-level overview of what is happening. Please let me know if anything (or rather what) is still unclear and needs better commenting.

I also added a bunch of unit tests for the different components. I would suggest enabling JIT-support for the CompareToSQlite tests for more coarse testing of the entire jitting pipeline on real queries.
I enabled this (for both Mac and Linux) on my machine with LLVM5 and all tests still pass.
With LLVM6 there are still some issues, which I believe Fabian is looking into.
Speaking of LLVM6: This branch is compatible with master and is thus ported to LLVM6. I took the modified `CloneFunction.cpp` / `InlineFunction.cpp` filed from Fabian's branch for that. All currently enabled test are happy with LLVM6.

I also had to add a `valgrind` suppression for the `JitRuntimePointer::is_valid` method, since this method is responsible for checking the validity of pointer values. So when testing this I deliberately feed invalid pointers to the method, which freaks out `valgrind`.

FabianWiebe commented on src/lib/logical_query_plan/jit_aware_lqp_translator.hpp:
```diff
@@ -1,7 +1,7 @@
 #pragma once
 
 #include "lqp_translator.hpp"
-#include "operators/jit_operator/operators/jit_compute.hpp"
```
In file `jit_aware_lqp_translator.cpp` the class `JitCompute` is used. So if the include is removed here, I would suggest to add it to to the file `jit_aware_lqp_translator.cpp`. Otherwise compilation is failing at least on my mac.

mrzzzrm commented:
If there is any way to test this, please add tests. It's much easier to read code from its usage downward then to just read it. ...and, obviously, tests test things.

mrks commented on src/test/operators/jit_operator/jit_aware_lqp_translator_test.cpp:
```diff
@@ -0,0 +1,387 @@
+  ASSERT_EQ(input_literals[1].tuple_value.is_nullable(), false);
+}
+
+TEST_F(JitAwareLQPTranslatorTest, SelectedColumnsAreOutputInCorrectOrder) {
```
Any reason that this isn't three independent tests?

jfrohnhofen replied:
They are now\

mrks commented on src/test/operators/jit_operator/jit_aware_lqp_translator_test.cpp:
```diff
@@ -0,0 +1,387 @@
+  ASSERT_EQ(a_gt_b->left_child()->expression_type(), ExpressionType::Column);
+  ASSERT_EQ(a_gt_b->right_child()->expression_type(), ExpressionType::Column);
+  ASSERT_TRUE(jit_read_tuples->find_input_column(a_gt_b->left_child()->result()) == ColumnID{0});
+  ASSERT_TRUE(jit_read_tuples->find_input_column(a_gt_b->right_child()->result()) == ColumnID{1});
```
@mrzzzrm - This could be shorter and easier, correct?

jfrohnhofen replied:
I thought about adding a `==` operator to `JitExpression`. However, this would be the only place where it would be used and in the test case we would still need to manually create the expression to compare to. That is less but noch much less code.
While the current version is quite verbose, it explicitly shows what is being checked instead of calling a "black box" `==` operator. I am happy to change it though if you want.

mrks commented on src/test/operators/jit_operator/specialization/jit_repository_test.cpp:
```diff
@@ -0,0 +1,69 @@
+  // Check that all defined methods in the class hierarchy are present in the bitcode repository.
+  // Virtual methods that are not implemented should cause a nullptr.
+  // See "src/test/llvm/virtual_methods.cpp" for the class hierarchy.
+  ASSERT_EQ(repository.get_function("_ZN4Base3fooEv"), nullptr);
```
Is there a good way to have actual signatures instead of mangled names here?

jfrohnhofen replied:
short answer: no
long answer: I am/was pretty annoyed by this as well and spend several days finding a solution for this a while back.
C++ name mangling is not part of LLVM itself but `libclang`. This would add another big dependency to Hyrise. And even withing clang the name mangling is not well isolated (there is no simple `get_mangled_name()` function, since the mangling relies on too much type information are requires a full compilation unit to be passed to `libclang`).
The solutions on stackoverflow either do implement their own name mangling with dodgy regular expressions or rely on generating code, calling `clang` and parsing the resulting bitcode string.

jfrohnhofen commented on exported_table_statistics.json:
```diff
@@ -0,0 +1 @@
+{"column_statistics":[{"data_type":"int","distinct_count":50.0999984741211,"max":100,"min":21,"null_value_ratio":0.300000011920929},{"data_type":"long","distinct_count":51.2000007629395,"max":101,"min":22,"null_value_ratio":0.400000005960464},{"data_type":"float","distinct_count":51.2999992370605,"max":1.00999999046326,"min":2.20000004768372,"null_value_ratio":0.5},{"data_type":"double","distinct ...
```
removed, but I did not change `.gitignore` since I feel this does not belong into this PR. Should I?

jfrohnhofen commented on src/lib/operators/jit_operator/jit_types.hpp:
```diff
@@ -135,6 +135,11 @@ class JitTupleValue {
 
+  bool operator==(const JitTupleValue& other) const {
+    return data_type() == other.data_type() && is_nullable() == other.is_nullable() &&
+           tuple_index() == other.tuple_index();
```
The method does not compare values, but data structures that are used to locate values in the runtime tuple. Equal `JitTupleValues` refer to the same value in a tuple. I added a comment to explain this.

jfrohnhofen commented on src/test/operators/jit_operator/specialization/jit_compiler_test.cpp:
```diff
@@ -0,0 +1,49 @@
+    llvm::InitializeNativeTargetAsmPrinter();
+
+    _context = std::make_shared<llvm::LLVMContext>();
+    _module = load_module("src/test/llvm/add.ll", *_context);
```
The bitcode is now generated automatically at compile time. I added a comment to the JIT section of `src/test/CMakeLists.txt` and added a comment to the test as well.

jfrohnhofen commented on src/lib/operators/jit_operator/specialization/llvm/InlineFunction.diff:
```diff
@@ -0,0 +1,37 @@
+14a15,18
+> // clang-format off
```
added readme

mrks commented on src/lib/operators/jit_operator/specialization/jit_code_specializer.hpp:
```diff
@@ -0,0 +1,70 @@
+ * The class has a single virtual method.
+ * A raw pointer can be cast to a pointer of this class to be used with RTTI.
+ */
+class JitRTTIHelper {
```
What is the problem with typeid?

jfrohnhofen replied:
I use typeid, but typeid already requires a typed pointer while I have a raw address that is computed by simulating bitcode instructions.
So I bet the pointer as an integer and then need to cast it to a pointer of some virtual class to make it usable with typeid.

mrks replied:
Can you store the type alongside with the pointer, similar to what happens with variants?

jfrohnhofen replied:
This does not really work, since we need to relate LLVM types (LLVMs internal representation of the types in the bitcode) to "real" runtime types.
More fundamentally, at some point we want to marry LLVM bitcode with runtime data. This only works because both originate from the same original source code. But while we "know" this, there is no type-safe way to accomplish this, since both bitcode and runtime objects are unrelated from the compilers POV. So even if we get rid of `reinterpret_cast`s there will always be unsafe operations, I think.

mrks commented:
It's great to see those many changes coming in over the last days. Could you let us know once this is stable?

jfrohnhofen commented on src/lib/operators/jit_operator/specialization/jit_code_specializer.cpp:
```diff
@@ -0,0 +1,310 @@
+      if (called_runtime_value && called_runtime_value->is_valid()) {
+        const auto vtable_index =
+            called_runtime_value->up().total_offset() / context.module->getDataLayout().getPointerSize();
+        const auto instance = reinterpret_cast<JitRTTIHelper*>(called_runtime_value->up().up().base().address());
```
Added comments

jfrohnhofen commented:
This can now be considered stable and is ready for another round of reviews.
I addressed all the feedback I got so far and added more documentation and comments to give a high-level overview of what is happening. Please let me know if anything (or rather what) is still unclear and needs better commenting.

I also added a bunch of unit tests for the different components. I would suggest enabling JIT-support for the `CompareToSQlite` tests for more coarse testing of the entire jitting pipeline on real queries.
I enabled this (for both Mac and Linux) on my machine with LLVM5 and all tests still pass.
With LLVM6 there are still some issues, which I believe Fabian is looking into.
Speaking of LLVM6: This branch is compatible with `master` and is thus ported to LLVM6. I took the modified `CloneFunction.cpp` / `InlineFunction.cpp` filed from Fabian's branch for that. All currently enabled test are happy with LLVM6.

I also had to add a `valgrind` suppression for the `JitRuntimePointer::is_valid` method, since this method is responsible for checking the validity of pointer values. So when testing this I deliberately feed invalid pointers to the method, which freaks out `valgrind`.

mrks commented:
@FabianWiebe Should we merge this now or should we wait for the LLVM 6 issue to be resolved?

FabianWiebe commented:
@mrks I would not wait till the issue is resolved as I am not sure how soon this will be.
